Rename cross compilation related defines

sdmaclea opened the pull request:
```
Rename cross compilation related defines

To drive explicit and consitent usage of the terms TARGET and HOST, and
to remove usage of reserved identifiers, rename several macro defines

  _ARM_                -> HOST_ARM       in src/coreclr
  _ARM64_              -> HOST_ARM64     in src/coreclr
  _AMD64_              -> HOST_AMD64     in src/coreclr
  _X86_                -> HOST_X86       in src/coreclr
  _HOST_UNIX_          -> HOST_UNIX      in src/coreclr
  _HOST_AMD64_         -> HOST_AMD64     in src/coreclr
  _HOST_ARM64_         -> HOST_ARM64     in src/coreclr
  _HOST_ARM_           -> HOST_ARM       in src/coreclr
  _HOST_X86_           -> HOST_X86       in src/coreclr
  DBG_TARGET_AMD64     -> TARGET_AMD64   in src/coreclr
  DBG_TARGET_ARM64     -> TARGET_ARM64   in src/coreclr
  DBG_TARGET_ARM       -> TARGET_ARM     in src/coreclr
  DBG_TARGET_X86       -> TARGET_X86     in src/coreclr
  _TARGET_AMD64_       -> TARGET_AMD64   in src/coreclr
  _TARGET_ARM_         -> TARGET_ARM     in src/coreclr
  _TARGET_ARM64_       -> TARGET_ARM64   in src/coreclr
  _TARGET_X86_         -> TARGET_X86     in src/coreclr
  _TARGET_ARMARCH_     -> TARGET_ARMARCH in src/coreclr
  _TARGET_XARCH_       -> TARGET_XARCH   in src/coreclr
  _HOST_64BIT_         -> HOST_64BIT     in src/coreclr
  BIT64                -> HOST_64BIT     in src/coreclr
  _TARGET_64BIT_       -> TARGET_64BIT   in src/coreclr
  DBG_TARGET_64BIT     -> TARGET_64BIT   in src/coreclr
  HOST_IS_WINDOWS_OS   -> HOST_WINDOWS   in src/coreclr
  PLATFORM_UNIX        -> HOST_UNIX      in src/coreclr/*.cmake
  PLATFORM_WINDOWS     -> HOST_WINDOWS   in src/coreclr/*.cmake
  _TARGET_MAC64        -> TARGET_OSX     in src/coreclr
  FEATURE_PAL          -> TARGET_UNIX    in src/coreclr
  _TARGET_UNIX_        -> TARGET_UNIX    in src/coreclr
  _TARGET_WINDOWS_     -> TARGET_WINDOWS in src/coreclr
  PLATFORM_UNIX        -> TARGET_UNIX    in src/coreclr

The first commit is purely a big search and replace.
The second commit is changes necessary to fix compilation issues.
The third commit is clang-format of the src/jit directory.
```

A future PR will rename some of the TARGET_* to HOST_* and vice-versa as needed.  This mostly a brute force search and replace.

jkotas commented on src/coreclr/clr.featuredefines.props:
```diff
@@ -60,7 +60,7 @@
         <DefineConstants Condition="'$(FeatureManagedEtw)' == 'true'">$(DefineConstants);FEATURE_MANAGED_ETW</DefineConstants>
         <DefineConstants Condition="'$(FeatureManagedEtwChannels)' == 'true'">$(DefineConstants);FEATURE_MANAGED_ETW_CHANNELS</DefineConstants>
-        <DefineConstants Condition="'$(FeaturePal)' == 'true'">$(DefineConstants);FEATURE_PAL</DefineConstants>
+        <DefineConstants Condition="'$(FeaturePal)' == 'true'">$(DefineConstants);TARGET_UNIX</DefineConstants>
```
I would be nice to delete `FeaturePal` as well (in future PR).

jkotas commented on src/coreclr/clrdefinitions.cmake:
```diff
@@ -179,8 +179,10 @@ if(FEATURE_MERGE_JIT_AND_ENGINE)
 add_compile_definitions($<$<NOT:$<BOOL:$<TARGET_PROPERTY:CROSSGEN_COMPONENT>>>:FEATURE_MULTICOREJIT>)
 if(CLR_CMAKE_TARGET_UNIX)
-  add_definitions(-DFEATURE_PAL)
-  add_definitions(-DFEATURE_PAL_ANSI)
```
`FEATURE_PAL_ANSI` should stay as `FEATURE_PAL_ANSI`

jkotas commented on src/coreclr/clr.featuredefines.props:
```diff
@@ -60,7 +60,7 @@
         <DefineConstants Condition="'$(FeatureManagedEtw)' == 'true'">$(DefineConstants);FEATURE_MANAGED_ETW</DefineConstants>
         <DefineConstants Condition="'$(FeatureManagedEtwChannels)' == 'true'">$(DefineConstants);FEATURE_MANAGED_ETW_CHANNELS</DefineConstants>
-        <DefineConstants Condition="'$(FeaturePal)' == 'true'">$(DefineConstants);FEATURE_PAL</DefineConstants>
+        <DefineConstants Condition="'$(FeaturePal)' == 'true'">$(DefineConstants);TARGET_UNIX</DefineConstants>
```
The CPU architecture defines have underscore around them. This one does not. It would be nice to be consistent (it will be a bug farm otherwise).

lpereira replied:
Not a good idea.  Identifiers starting with underscore are reserved for the standard, at least in C.  It's undefined behavior to define them yourself.  (The CPU architecture defines are like that because they're provided by the compiler.)

sdmaclea replied:
@lperiera For consistency it makes sense.  We use that style a lot currently.  If we want to change it should be a separate PR.

For the Preprocessor it is common practice to make file #ifndef to use a form `_Name_H_` or something like that.  Certainly we could collide with a compile #define, but it is unlikely.
 

```
$ git grep add_definitions | grep TARGET_
clrdefinitions.cmake:    add_definitions(-DDBG_TARGET_AMD64_UNIX)
clrdefinitions.cmake:  add_definitions(-D_TARGET_AMD64_)
clrdefinitions.cmake:  add_definitions(-D_TARGET_64BIT_)
clrdefinitions.cmake:  add_definitions(-DDBG_TARGET_64BIT)
clrdefinitions.cmake:  add_definitions(-DDBG_TARGET_AMD64)
clrdefinitions.cmake:    add_definitions(-DDBG_TARGET_ARM64_UNIX)
clrdefinitions.cmake:  add_definitions(-D_TARGET_ARM64_)
clrdefinitions.cmake:  add_definitions(-D_TARGET_64BIT_)
clrdefinitions.cmake:  add_definitions(-DDBG_TARGET_64BIT)
clrdefinitions.cmake:  add_definitions(-DDBG_TARGET_ARM64)
clrdefinitions.cmake:    add_definitions(-DDBG_TARGET_ARM_UNIX)
clrdefinitions.cmake:  add_definitions(-D_TARGET_ARM_)
clrdefinitions.cmake:  add_definitions(-DDBG_TARGET_32BIT)
clrdefinitions.cmake:  add_definitions(-DDBG_TARGET_ARM)
clrdefinitions.cmake:  add_definitions(-D_TARGET_X86_)
clrdefinitions.cmake:  add_definitions(-DDBG_TARGET_32BIT)
clrdefinitions.cmake:  add_definitions(-DDBG_TARGET_X86)
clrdefinitions.cmake:  add_definitions(-D_TARGET_UNIX_)
clrdefinitions.cmake:  add_definitions(-D_TARGET_WINDOWS_)
src/debug/debug-pal/CMakeLists.txt:  add_definitions(-D_TARGET_UNIX_)
src/debug/debug-pal/CMakeLists.txt:  add_definitions(-D_TARGET_WINDOWS_)
src/gcinfo/gcinfo_arm/CMakeLists.txt:add_definitions(-D_TARGET_ARM_)
src/gcinfo/gcinfo_arm64/CMakeLists.txt:add_definitions(-D_TARGET_ARM64_)
src/jit/armelnonjit/CMakeLists.txt:    add_definitions(-D_TARGET_ARM_)
src/jit/armelnonjit/CMakeLists.txt:    add_definitions(-D_TARGET_ARM64_)
src/jit/protononjit/CMakeLists.txt:    add_definitions(-D_TARGET_ARM_)
src/jit/protononjit/CMakeLists.txt:    add_definitions(-D_TARGET_ARM64_)
src/pal/src/CMakeLists.txt:  add_definitions(-D_TARGET_MAC64)
src/pal/src/CMakeLists.txt:add_definitions(-D_TARGET_UNIX_)
src/pal/tests/CMakeLists.txt:add_definitions(-D_TARGET_UNIX_)
tests/src/Interop/PInvoke/Generics/CMakeLists.txt:  add_definitions(-D_TARGET_X86_)
tests/src/Interop/PInvoke/Generics/CMakeLists.txt:  add_definitions(-D_TARGET_XARCH_)
tests/src/Interop/PInvoke/Generics/CMakeLists.txt:  add_definitions(-D_TARGET_AMD64_)
tests/src/Interop/PInvoke/Generics/CMakeLists.txt:  add_definitions(-D_TARGET_XARCH_)
tests/src/Interop/PInvoke/Generics/CMakeLists.txt:  add_definitions(-D_TARGET_ARM_)
tests/src/Interop/PInvoke/Generics/CMakeLists.txt:  add_definitions(-D_TARGET_ARMARCH_)
tests/src/Interop/PInvoke/Generics/CMakeLists.txt:  add_definitions(-D_TARGET_ARM64_)
tests/src/Interop/PInvoke/Generics/CMakeLists.txt:  add_definitions(-D_TARGET_ARMARCH_)
```

jkotas replied:
I am fine with either style - `_TARGET_XXX_` or `TARGET_XXX` - as long as it is used for all `XXX`.

lpereira commented on src/coreclr/clrdefinitions.cmake:
```diff
@@ -179,8 +179,10 @@ if(FEATURE_MERGE_JIT_AND_ENGINE)
 add_compile_definitions($<$<NOT:$<BOOL:$<TARGET_PROPERTY:CROSSGEN_COMPONENT>>>:FEATURE_MULTICOREJIT>)
 if(CLR_CMAKE_TARGET_UNIX)
-  add_definitions(-DFEATURE_PAL)
+  add_definitions(-D_TARGET_UNIX_)
```
Please see my comment about identifiers begining with an underscore. (They're UB in C.)

sdmaclea replied:
@lpereira I revised the patch try to fix many of the identifiers with an underscore.

jkotas commented:
> `BIT64               -> HOST_64BIT`

We have choose `TARGET_` as the default in all other place. Why use `HOST_` as the default for this one?

sdmaclea commented:
> > `BIT64               -> HOST_64BIT`
> 
> We have choose `TARGET_` as the default in all other place. Why use `HOST_` as the default for this one?

It was currently set based on the host architecture.  It looks wrong in most places, but I wanted to make this mostly a non-functional change.  Given the JIT supports crossbitness builds, I didn't want to mess with this here.

I also switched the PLATFORM_UNIX to map to HOST_UNIX on the similar rationale.

sdmaclea commented:
I certainly want to minimize `HOST_XXX` instances.  They seem problematic outside the PAL.  My guess is they are only likely to be right in JIT and zapper.  I will definitely audit them.

sdmaclea commented:
I had originally been doing the bulk edit replaces in my editor.  Turns out my editor was skipping some files for some reason....

Anyways I decided it was less error prone to script this.  So I wrote a bash/git/sed script to replace a single editor within a git subtree.  Then I wrote a script for all the bulk edits.

I kept the commits more separate to ease review burden.  The bulk search replace are separate commits.  

The manual edits are separate commits.

I replaced the BIT64 rename to TARGET_BIT64 except in cmake, zap and jit (which had no references).

sdmaclea committed:
Miscellaneous manual fixes to fix compilation

Remove unused defines

Remove BIT32
Remove DBG_TARGET_AMD64_UNIX
Remove DBG_TARGET_ARM64_UNIX
Remove DBG_TARGET_ARM_UNIX
Remove DBG_TARGET_32BIT

Fixes for HOST_<arch> rename

Move TARGET_<Arch> and TARGET_<bit>

Move from clrdefinitions.cmake
to configurecompiler.cmake so it is used globally.

More jit.h

sdmaclea committed:
Rename identifiers to be more consistent

This patch was a machine generated search and replace of
identifiers.  The replacement was done in the order listed
for the path specified.

A regex was used to find the identifiers.

s/(^|[^A-CE-Za-z0-9_])${find}($|[^A-Za-z0-9_])

The 'D' character was specifically allowed a s a prefix
to catch add_definitions(-D${find}).

The patch also reverts all managed code changes after replacement.

    _ARM_                -> HOST_ARM       in src/coreclr
    _ARM64_              -> HOST_ARM64     in src/coreclr
    _AMD64_              -> HOST_AMD64     in src/coreclr
    _X86_                -> HOST_X86       in src/coreclr
    _HOST_UNIX_          -> HOST_UNIX      in src/coreclr
    _HOST_AMD64_         -> HOST_AMD64     in src/coreclr
    _HOST_ARM64_         -> HOST_ARM64     in src/coreclr
    _HOST_ARM_           -> HOST_ARM       in src/coreclr
    _HOST_X86_           -> HOST_X86       in src/coreclr
    DBG_TARGET_AMD64     -> TARGET_AMD64   in src/coreclr
    DBG_TARGET_ARM64     -> TARGET_ARM64   in src/coreclr
    DBG_TARGET_ARM       -> TARGET_ARM     in src/coreclr
    DBG_TARGET_X86       -> TARGET_X86     in src/coreclr
    _TARGET_AMD64_       -> TARGET_AMD64   in src/coreclr
    _TARGET_ARM_         -> TARGET_ARM     in src/coreclr
    _TARGET_ARM64_       -> TARGET_ARM64   in src/coreclr
    _TARGET_X86_         -> TARGET_X86     in src/coreclr
    _TARGET_ARMARCH_     -> TARGET_ARMARCH in src/coreclr
    _TARGET_XARCH_       -> TARGET_XARCH   in src/coreclr
    _HOST_64BIT_         -> HOST_64BIT     in src/coreclr
    BIT64                -> HOST_64BIT     in src/coreclr
    _TARGET_64BIT_       -> TARGET_64BIT   in src/coreclr
    DBG_TARGET_64BIT     -> TARGET_64BIT   in src/coreclr
    HOST_IS_WINDOWS_OS   -> HOST_WINDOWS   in src/coreclr
    PLATFORM_UNIX        -> HOST_UNIX      in src/coreclr/*.cmake
    PLATFORM_WINDOWS     -> HOST_WINDOWS   in src/coreclr/*.cmake
    _TARGET_MAC64        -> TARGET_OSX     in src/coreclr
    FEATURE_PAL          -> TARGET_UNIX    in src/coreclr
    _TARGET_UNIX_        -> TARGET_UNIX    in src/coreclr
    _TARGET_WINDOWS_     -> TARGET_WINDOWS in src/coreclr
    PLATFORM_UNIX        -> TARGET_UNIX    in src/coreclr
    PLATFORM_WINDOWS     -> TARGET_WINDOWS in src/coreclr

sdmaclea committed:
Whitespace by clang-format

BruceForstall commented:
fwiw, this change deserved to be (and still deserves to be) much more widely before being merged. Anyone with in-progress changes that added architecture-specific code (for example), then rebased, might not get the behavior they expect because previously-working `#ifdef _TARGET_ARM64_` would now be dead code.

@dotnet/jit-contrib fyi: many global `#ifdefs` have changed names.

Also, can we change `TARGET_OSX` to `TARGET_MACOS`? Apple replaced the name `OSX` with `macOS` in 2016.

jkotas commented:
> Also, can we change TARGET_OSX to TARGET_MACOS? 

There is discussion about it here: https://github.com/dotnet/runtime/pull/31659#issuecomment-581614248 . 

We have OSX name in the number of places today that we cannot change (e.g. https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.osplatform.osx?view=netframework-4.8).

If we want to change it to something, I would vote for DARWIN. DARWIN should be immune to marketing whims.

AndyAyersMS commented:
Can we make the old names cause errors for some interim period, so it's more obvious they won't work anymore? Otherwise stuff may just silently get dropped.

CarolEidt commented:
I filed #31732 to clean up some inconsistencies in the JIT

sdmaclea commented:
@dotnet/jit-contrib My apologies for not socializing this earlier.  I had been concerned that my PR would shear with respect to master.  That somehow I would have missed a rename....  I checked immediately after I merged so that I could prevent an outage.  I didn't really anticipate the possibility that other PRs might go in after w/o having merge conflicts with this one....

@AndyAyersMS I am not sure how to make use of an undefined macro cause a build break.  

Something like 
```C++
#define _TARGET_ARM_ static_assert(BadDefine != BadDefine, "_TARGET_ARM_ macro is obsolete.  Use TARGET_ARM")
```

Might work in some circumstances, but not in all. I don't think it would work in the most common cases.  Particularly:

```C++
#ifdef _TARGET_ARM_
```
```C++
#if defined(_TARGET_ARM_)
```

erozenfeld commented:
Will 
`#pragma deprecated("_TARGET_ARM_")`
work?
From https://docs.microsoft.com/en-us/cpp/preprocessor/deprecated-c-cpp?view=vs-2019 
`You can deprecate macro names. Place the macro name in quotes or else macro expansion will occur.`

lpereira commented:
For non-MSVC compilers, that `#pragma` won't work.  In C (from C99 onwards at least), there's `_Pragma()`, and using it like so works in GCC and (probably) Clang:

```C
#define WARN(msg) WARN2(GCC warning msg)
#define WARN2(s) _Pragma(#s)
#define DEPRECATE_MACRO WARN("Macro is deprecated")
```

And then use it like so:

```C
#define _TARGET_ARM_ DEPRECATE_MACRO
```

I don't know, however, if there's something similar for C++ or even MSVC, though.

janvorli commented:
While this mechanism works with clang, it doesn't help our purpose. The problem is that it only warns for `#if _TARGET_ARM_`, but not for `#ifdef _TARGET_ARM_`. The latter is used in all the cases we have. Unfortunately, I don't see any way to create a warning for the #ifdef usage case.
